### ISSUE TITLE
Skill Model, Basic usage

### ISSUE THREAD

**Issue by erlebach** [biaslab/ReactiveMP.jl#102]
I am trying to apply ReactiveMP to Winn and Bishop's skill model. I am starting ultra simple and will build up. I just learned to do this in ForneyLab with help from the developers. I hope to do the same here. What I have so far, modeling my efforts to mimic the Coin Flip problem since my model is discrete. Of course I get an error. 

``` Julia
using Rocket, GraphPPL, ReactiveMP, Distributions, Random

@model function toy3(n)
    y = datavar(n)  # 3 questions
    
    csharp ~ Bernoulli(0.5)
    sql ~ Bernoulli(0.5)
    
    isCorrect1 ~ Bernoulli(Addnoise(csharp))
    isCorrect2 ~ Bernoulli(Addnoise(sql))
    
    return y, isCorrect1, isCorrect2
end
```

``` Julia
dataset = [1., 0.]

result = inference(
    model = Model(toy3, length(dataset)),
    data  = (y = dataset, )
)
```

Error: 
``` Julia
lid type object was expected but '2' has been found

Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:33
  [2] ensure_type(x::Int64)
    @ GraphPPL ~/.julia/packages/GraphPPL/KNm3q/src/model.jl:22
  [3] macro expansion
    @ ./In[56]:4 [inlined]
  [4] macro expansion
    @ ~/.julia/packages/GraphPPL/KNm3q/src/model.jl:347 [inlined]
  [5] create_model(::Type{toy3}, toy3constraints_in#451::Nothing, toy3meta_in#453::Nothing, toy3options_in#455::NamedTuple{(), Tuple{}}, n::Int64)
    @ Main ~/.julia/packages/GraphPPL/KNm3q/src/backends/reactivemp.jl:73
  [6] create_model(generator::ReactiveMP.ModelGenerator{toy3, Tuple{Int64}, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, Nothing, Nothing, Nothing}, constraints::Nothing, meta::Nothing, options::NamedTuple{(), Tuple{}})
    @ ReactiveMP ~/src/2022/ReactiveMP.jl/src/model.jl:42
  [7] inference(; model::ReactiveMP.ModelGenerator{toy3, Tuple{Int64}, Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, Nothing, Nothing, Nothing}, data::NamedTuple{(:y,), Tuple{Vector{Float64}}}, initmarginals::Nothing, initmessages::Nothing, constraints::Nothing, meta::Nothing, options::NamedTuple{(), Tuple{}}, returnvars::Nothing, iterations::Int64, free_energy::Bool, showprogress::Bool, callbacks::Nothing, warn::Bool)
    @ ReactiveMP ~/src/2022/ReactiveMP.jl/src/inference.jl:274
  [8] top-level scope
    @ In[57]:3
  [9] eval
    @ ./boot.jl:373 [inlined]
 [10] include_string(mapexpr::typeof(REPL.softscope), mod::Module, code::String, filename::String)
    @ Base ./loading.jl:1196
```

Obviously, there is some basic concept I do not understand, although I thought I did. What could have gone amiss? Thanks for any input!

    Gordon

**Comment by bvdmitri** [comment]
`datavar()` function expects `Type` as its first argument. Check our [documentation](https://biaslab.github.io/ReactiveMP.jl/stable/man/model-specification/#user-guide-model-specification-data-variables) regarding model specification.

**Comment by erlebach** [comment]
Thanks! I fixed this problem (oversight on my part). Now I get another error. Is there a way to see the source code generated for the model?   I define a function `AddNoise`, but noticed that if I define the function outside the model macro `@model`, and `undefined function` error occurs. So I define the function inside the model. Here is my current model (which generates an error on compilation): 

``` Julia
# AddNoise(hasSkill) = if hasSkill == 1 return 0.9 else return 0.2 end

@model function toy3(n)
    AddNoise(hasSkill) = if hasSkill == 1 return 0.9 else return 0.2 end

    y = datavar(Float64, n)  # 3 questions
    
    csharp ~ Bernoulli(0.5)
    sql ~ Bernoulli(0.5)
    
    isCorrect1 ~ Bernoulli(Addnoise(csharp))
    isCorrect2 ~ Bernoulli(Addnoise(sql))
    
    return y, isCorrect1, isCorrect2
end
```
and the full error message: 
``` Julia
LoadError: Variable identificator can be in form of a single symbol (x ~ ...) or indexing expression (x[i] ~ ...)
in expression starting at /Users/erlebach/.julia/packages/GraphPPL/KNm3q/src/model.jl:178

Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:33
  [2] parse_varexpr(varexpr::Expr)
    @ GraphPPL ~/.julia/packages/GraphPPL/KNm3q/src/model.jl:49
  [3] (::GraphPPL.var"#42#55"{Set{Symbol}})(expression::Expr)
    @ GraphPPL ~/.julia/packages/GraphPPL/KNm3q/src/model.jl:282
  [4] walk
    @ ~/.julia/packages/MacroTools/PP9IQ/src/utils.jl:112 [inlined]
  [5] postwalk
    @ ~/.julia/packages/MacroTools/PP9IQ/src/utils.jl:122 [inlined]
  [6] (::MacroTools.var"#21#22"{GraphPPL.var"#42#55"{Set{Symbol}}})(x::Expr)
    @ MacroTools ~/.julia/packages/MacroTools/PP9IQ/src/utils.jl:122
  [7] iterate
    @ ./generator.jl:47 [inlined]
  [8] collect_to!(dest::Vector{LineNumberNode}, itr::Base.Generator{Vector{Any}, MacroTools.var"#21#22"{GraphPPL.var"#42#55"{Set{Symbol}}}}, offs::Int64, st::Int64)
    @ Base ./array.jl:782
  [9] collect_to_with_first!(dest::Vector{LineNumberNode}, v1::LineNumberNode, itr::Base.Generator{Vector{Any}, MacroTools.var"#21#22"{GraphPPL.var"#42#55"{Set{Symbol}}}}, st::Int64)
    @ Base ./array.jl:760
 [10] _collect(c::Vector{Any}, itr::Base.Generator{Vector{Any}, MacroTools.var"#21#22"{GraphPPL.var"#42#55"{Set{Symbol}}}}, #unused#::Base.EltypeUnknown, isz::Base.HasShape{1})
    @ Base ./array.jl:754
 [11] collect_similar(cont::Vector{Any}, itr::Base.Generator{Vector{Any}, MacroTools.var"#21#22"{GraphPPL.var"#42#55"{Set{Symbol}}}})
    @ Base ./array.jl:653
 [12] map(f::Function, A::Vector{Any})
    @ Base ./abstractarray.jl:2849
 [13] walk(x::Expr, inner::Function, outer::GraphPPL.var"#42#55"{Set{Symbol}})
    @ MacroTools ~/.julia/packages/MacroTools/PP9IQ/src/utils.jl:112
 [14] postwalk(f::Function, x::Expr)
    @ MacroTools ~/.julia/packages/MacroTools/PP9IQ/src/utils.jl:122
 [15] generate_model_expression(backend::ReactiveMPBackend, model_options::Expr, model_specification::Expr)
    @ GraphPPL ~/.julia/packages/GraphPPL/KNm3q/src/model.jl:279
 [16] var"@model"(__source__::LineNumberNode, __module__::Module, model_options::Any, model_specification::Any)
    @ GraphPPL ~/.julia/packages/GraphPPL/KNm3q/src/model.jl:182
 [17] eval
    @ ./boot.jl:373 [inlined]
 [18] include_string(mapexpr::typeof(REPL.softscope), mod::Module, code::String, filename::String)
    @ Base ./loading.jl:1196
```

**Comment by bvdmitri** [comment]
ReactiveMP.jl does not generate source-code in the same way as ForneyLab.jl. Regarding this error: You cannot define another functions defined in `@model` macro (this is an issue we can fix in the future). You can define a function outside of the `@model` macro and use it as usual. I suppose you just misspelled (`AddNoise` in the definition and `Addnoise` in the model specifiction, note capital `N`) your function name - so Julia shows `undefined function` error. 

I would warn you, however, that ReactiveMP.jl does not support arbitrary functions within model specification. You can follow this [guide](https://github.com/biaslab/ReactiveMP.jl/blob/master/demo/Custom%20nonlinear%20node.ipynb) to check how to create custom nonlinear nodes.

**Comment by erlebach** [comment]
Thanks! Problem fixed. I am wondering:  have you implemented the ability to go back and forth between weighted samples and distributions as is implemented in ForneyLab with the Extended Variational Message Passing? I tried that first but was blocked by ForneyLab's inability to handle cyclic graphs.

**Comment by bvdmitri** [comment]
We are working hard to make it work, but currently Extended VMP methods under development. It is hard to make any estimates, but we are hoping to make it stable this year.

**Comment by erlebach** [comment]
Thanks for the info! This is difficult stuff!

**Comment by bvdmitri** [comment]
It is not easy, indeed. Do not hesitate to ask questions or if you find another issue. Can I close this issue?

**Comment by erlebach** [comment]
Thanks. Yes, please close this issue.

**Comment by albertpod** [comment]
@erlebach, we thank you for trying out `ReactiveMP.jl`
Your feedback will help us to improve our framework. 
Soon we will support the type of models you work with.
